src/bank02.asm

```diff
@@ -8130,29 +8126,194 @@ titleScreenLicenseText:
-    TXT  "it gives eternal<00>"                        ;; 02:7f3e ?????????????????
-    TXT  "power to the one<00>"                        ;; 02:7f4f ?????????????????
-    TXT  " who touched it.<00>"                        ;; 02:7f60 ?????????????????
-    TXT  "<00>"                                        ;; 02:7f71 ?
-    TXT  "  Dark Lord was<00>"                         ;; 02:7f72 ????????????????
-    TXT  "  trying to find<00>"                        ;; 02:7f82 ????????????????.
-    TXT  "  the way to the<00>"                        ;; 02:7f93 ...........??????
-    TXT  " Tree of Mana to<00>"                        ;; 02:7fa4 ?????????????????
-    TXT  "  get the mighty<00>"                        ;; 02:7fb5 ?????????????????
-    TXT  " power to conquer<00>"                       ;; 02:7fc6 ??????????????????
-    TXT  "    the world.<00>"                          ;; 02:7fd8 ???????????????
-    TXT  "<00>"                                        ;; 02:7fe7 ?
-    db   $00, $00, $00, $00, $00, $00, $00, $00        ;; 02:7fe8 ????????
-    db   $00, $00, $00, $00, $01, $ff, $ff, $ff        ;; 02:7ff0 ????????
-    db   $ff, $ff, $ff, $ff, $ff, $ff, $ff, $ff        ;; 02:7ff8 ????????
+    TXT  "<00>"                                        ;; 02:7fd8 ?
+    db   $01                                           ;; 02:7fd9 ?
+
+; Draws left aligned number HL as modified tiles at WRAM offset E
+; D holds the drawing mode in the first nibble (0 for shift, 1 for shift/swap)
+; D holds number of max digits in the second nibble
+drawLeftAlignedNumberInWRAM:
+    ld   B, $00 ; digit counter
+    ld   C, D ; use C to hold mode/max-digit-count
+
+    ; Housekeeping, get a copy of proper WRAM starting
+    ;  location on the stack for later use
+    push HL
+    ld   HL, wHPMPTileBuffer
+    ld   D, B
+    add  HL, DE
+    ld   D, H
+    ld   E, L
+    pop  HL
+    push BC
+    push DE
+
+    ; Load up the digits on the stack
+.divmod_and_store_number:
+    ld   A, $0a
+    push BC
+    push DE
+    call divMod
+    pop  DE
+    pop  BC
+    push AF
+    inc  B
+    ld   A, H
+    or   A, L
+    jr   NZ, .divmod_and_store_number
+
+    ; Reduce second nibble of C by num digits found
+    ld   A, C
+    sub  A, B
+    ld   C, A
+
+    ; Increment DE if in swap mode
+    ld   H, D
+    ld   L, E
+    ld   D, $00
+    and  A, $f0
+    ld   E, A
+    add  HL, DE
+    ld   D, H
+    ld   E, L
+
+    ; Stack holds B digits to pop, transfer them
+ .grab_digit_and_transfer:
+    ld   HL, $6800
+    pop  AF ; grab digit
+    swap A
+    push DE
+    ld   D, $00
+    ld   E, A
+    add  HL, DE
+    pop  DE
+    ld   A, $08
+    push BC
+    ld   B, $10
+    call copyShiftBankAfromHLtoDE
+    pop  BC
+    dec  B
+    jr   NZ, .grab_digit_and_transfer
+
+    ; Clear remaining tiles
+    ld   A, C
+    and  A, $0f
+    jr   Z, .check_swap
+
+    ld   H, D
+    ld   L, E
+    ld   B, A
+    xor  A, A
+ .clear_remaining_tiles:
+    call memsetTileWithA
```

Comment:
There are a few ways this could be dealt with instead of adding a new function.

Easiest would be to just use `fillMemory` (which already has a check for `B == 0`), or just add a simple loop here.

A smarter way would be to leave the tiles dirty and request a transfer of the canonical blank tile `gfxBlankTiles08` instead. That would mean coordination between `drawLeftAlignedNumberInWRAM` and `requestVRAMStatusBarTransfer`, or merging the two.

Least resource intensive would be to use `storeTileAatWindowPositionDE` to change the tiles to `7f` (and change them back each call of `drawLeftAlignedNumberInWRAM`, but that might complicate the code, or lead to partial updates. I think given the code I've looked at so far along with your testing results, this option might be best.

What do you think of these options? At the moment I'm sensitive to the size of this function, but it also seems like it's probably not a useful optimization?

Reply:
As for the first option, the memsetTileWithA routine is much more efficient than fillMemory (36 cycles vs 135 cycles per tile).

I think the third option squarely beats the second option. Two issues: 
1) if I leave the extra tiles in WRAM with junk, it'll be copied back across the tile boundary as part of the swap routine later unless I add special logic in the loop. I think that'd look like running the loop one time less than typical, then doing special logic at the end of the loop. Not terrible.
2) I'd have to communicate between the functions. Thankfully they're right after eachother, so I can effectively have return codes in the registers. Doable.

Is it worth the optimization? Not really. If late game we're using all the tiles most of the time anyway, why bother optimizing for the early game?

Reply:
For `fillMemory` (or an in-line loop) I was thinking the per-tile loop could be replaced with a single counter. That would save some time in setup and calling for each tile. I'm seeing the current solution as 46 cycles (including the call, dec, and jr), and a single loop would be 96 cycles. Still worse, but not as bad. 

That said, can this loop also clear the tile used in `swapBackHalfTiles`? If it could easily (and I think it can, but I haven't checked), that would bring up the question of whether to inline `memsetTileWithA` here.

Reply:
I think it'd be ok to either:
1) Move memsetTileWithA to bank00 and call it "fillTile" to match "fillMemory".
2) Just use fillMemory, as much as it pains my inner optimizer.

What do you think?

Reply:
Looks like we commented at about the same time. I didn't see yours before posting mine.

I moved the first tile memset call into the drawLeft... function. I don't really think sticking it in the same loop as clearing the remainder tiles makes sense. That seems like a lot of extra logic for inlining that function.

As for the single byte loop (I think that's what you meant?), I think I'd rather just use fillMemory even though it's less efficient. That way, when I come through with the optimization hammer after improving the rendering system, I'll easily spot the inefficiency. Want me to change it?